### ISSUE TITLE
[no-unnecessary-type-assertion] False positive with value that can have the value null

### ISSUE THREAD

**Issue by mdouglass** [typescript-eslint/typescript-eslint#1199]
Ok, I know I already tried to open this as #1195, but I totally borked that report. Hopefully, I'm not doing it again 🤞, the issue was with no-unnecessary-type-assertion all along.  

**Repro**
```JSON
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": { "project": "./tsconfig.json", "sourceType": "module" },
  "plugins": ["@typescript-eslint"],
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

```TS
function testFunction(_param: string | undefined): void { /* noop */ }
const value = 'test' as string | null | undefined
testFunction(value!)
```

**Expected Result**
No errors

**Actual Result**
  3:14  error  This assertion is unnecessary since the receiver accepts the original type of the expression  @typescript-eslint/no-unnecessary-type-assertion

**Additional Info**
With the `!` this compiles, but if the `!` is removed, then you get a compiler error:
```
Argument of type 'string | null | undefined' is not assignable to parameter of type 'string | undefined'.
  Type 'null' is not assignable to type 'string | undefined'.ts(2345)
```

**Versions**

| package | version |
| ---------------------------------- | ------- |
| @typescript-eslint/eslint-plugin | 2.6.1 |
| @typescript-eslint/parser | 2.6.1 |
| TypeScript | 3.7.2 |
| ESLint | 6.6.0 |
| node | 13.1.0 |
| npm | 6.12.1 |

**Comment by bradzacher** [comment]
This report makes more sense :)

This is actually something I found a while ago but never got around to fixing.
I'll see if I can put up a PR today.

**Comment by mdouglass** [comment]
Thank goodness :) 
And thanks for taking a look at it!
